proposals/0000-slurp.rst

```diff
@@ -0,0 +1,454 @@
+-  A common format for package metadata (i.e. Cabal) and a common way to
+   build packages.
+-  A standard API for performing (1), (2) and (3) above.
+
+Given a central registry to locate packages (SLURP), any build tool or
+package infrastructure can query which server a package is hosted on.
+Then, provided that server implements a standard API, the build tool can
+query available versions and download source distributions.
+
+4 The SLURP API
+---------------
+
+The central registry, SLURP, is a new service. Its purpose is to allow
+clients claim a package name, to list all available packages, and to
+find where a package is hosted. It implements the following API:
+
+-  ``GET /packages`` - returns a list of all open source packages in the
+   Haskell universe.
+-  ``PUT /package/:pkgname`` - register ownership of a new package name.
+   Map that name to some URL that you own. Example:
+   ``{"name": "mypackage", "location": "https://myserver.com/package/mypackage" }``
+-  ``GET /package/:pkgname`` - redirects to URL provided by ``PUT``.
+
+No authentication or user accounts required to make requests to either
+of those endpoints.
+
+Example package list returned as the response of a ``GET /packages``
+request:
+::
+
+    { "packages":
+      [ {"name": "bytestring", "location": "https://hackage.haskell.org/package/bytestring"}
+      , {"name": "text", "location": "https://hackage.haskell.org/package/text"}
+      , {"name": "Cabal", "location": "https://hackage.haskell.org/package/Cabal"}
+      , {"name": "conduit", "location": "https://stackage.org/package/conduit" }
+      , {"name": "yesod", "location": "https://stackage.org/package/yesod" }
+      , {"name": "mypackage", "location": "https://myserver.com/package/mypackage" }
+      ...
+      ]
+    }
+
+You can think of SLURP as an *authoritative* URL-shortening service for
+packages. Authoritative means that which package is generally accepted
+to have which name depends on the content of the SLURP registry and only
+that.
+
+5 The package API
+-----------------
+
+The target location for each registered package must support at least
+the following subset of the Hackage API:
+
+-  ``GET /package/:pkgname/preferred`` returns a JSON structure listing
+   all versions.
+-  ``GET /package/:pkgname/:tarball.tar.gz`` where tarball is a
+   name-version pair.
+
+We choose to define the package API as a subset of the current Hackage
+API for backwards compatibility: it means Hackage is a valid package
+hosting server from Day 1, without change.
+
+We put forward no requirement on the guarantees provided by package
+hosting services. For example, we do not impose that getting a
+particular package tarball for (say) ``pkg-3.5.3`` always yields the
+same result each time. (In Hackage, for example, the presence of
+revisions may mean it will not.) Immutability, or lack thereof, is an
+example of hosting-service specific policy that does not concern SLURP.
```

Comment:
I have serious doubts about this aspect.

Downstream packagers like Debian would have serious trouble if downloading `pkg-3.5.3.tar.gz` would yield different results later on. In general, it would cause huge amounts of confusion.

Also  note that (AFAIK) hackage does not actually ship mutable tarballs. It ships updated `.cabal` files out-of-band (in the `00-index.tar`), and `cabal-install` upon unpacking replaces the `.cabal` file.

Is this paragraph motivated by a real use case where hackage/stackage/whatever plan to actually ship tarballs that vary over time? Or is it in to cater for the Hackage revision feature, written with maybe a slightly wrong idea about how it actually works?

Could this paragraph be replaced by something like “We impose that getting a particular package tarball always yields the same results. Hosting services that plan to modify the package in some way should deliver this modification out-of-band in a way that is specifically understood by the corresponding client.”?